contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,699 @@
+     * @return              This address to accept, a different address to ask that contract
+     */
+    function verifyLoanOffering(
+        address[10] addresses,
+        uint256[7] values256,
+        uint32[4] values32,
+        bytes32 positionId
+    )
+        external
+        onlyMargin
+        nonReentrant
+        returns (address)
+    {
+        LoanOffering memory loanOffering = parseLoanOffering(addresses, values256, values32);
+
+        /* CHECK POSITIONID */
+        require(positionId == POSITION_ID);
+
+        /* CHECK ADDRESSES */
+        require(loanOffering.owedToken == OWED_TOKEN);
+        require(loanOffering.heldToken == HELD_TOKEN);
+        require(loanOffering.payer == address(this));
+        // no need to require anything about loanOffering.signer
+        require(loanOffering.owner == address(this));
+        // no need to require anything about loanOffering.taker
+        // no need to require anything about loanOffering.positionOwner
+        // no need to require anything about loanOffering.feeRecipient
+        // no need to require anything about loanOffering.lenderFeeToken
+        // no need to require anything about loanOffering.takerFeeToken
+
+        /* CHECK VALUES256 */
+        // no need to require anything about loanOffering.maximumAmount
+        // no need to require anything about loanOffering.minimumAmount
+        // no need to require anything about loanOffering.minimumHeldToken
+        require(loanOffering.lenderFee == 0);
+        // no need to require anything about loanOffering.takerFee
+        // no need to require anything about loanOffering.expirationTimestamp
+        // no need to require anything about loanOffering.salt
+
+        /* CHECK VALUES32 */
+        // no need to require anything about loanOffering.callTimeLimit
+        // no need to require anything about loanOffering.maxDuration
+        // no need to require anything about loanOffering.interestRate
+        // no need to require anything about loanOffering.interestPeriod
+
+        return address(this);
+    }
+
+    /**
+     * Called by the Margin contract when anyone transfers ownership of a loan to this contract.
+     * This function initializes this contract and returns this address to indicate to Margin
+     * that it is willing to take ownership of the loan.
+     *
+     * @param  from        (unused)
+     * @param  positionId  Unique ID of the position
+     * @return             This address on success, throw otherwise
+     */
+    function receiveLoanOwnership(
+        address from,
+        bytes32 positionId
+    )
+        external
+        onlyMargin
+        onlyPosition(positionId)
+        returns (address)
+    {
+        MarginCommon.Position memory position = MarginHelper.getPosition(DYDX_MARGIN, POSITION_ID);
+
+        assert(position.principal > 0);
+        assert(position.owedToken == OWED_TOKEN);
+        assert(position.heldToken == HELD_TOKEN);
+
```

Comment:
maybe also `assert(principal.total == 0)` or something to ensure that this contract hasn't been initialized yet. I know it should never be possible with the base protocol, but good to have